craft_application/models/platforms.py

```diff
@@ -0,0 +1,226 @@
+# This file is part of craft-application.
+#
+# Copyright 2023-2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Models that describe platforms."""
+
+import enum
+from collections.abc import Iterable, Mapping
+from typing import ClassVar, get_args
+
+import craft_platforms
+import pydantic
+from pydantic_core import core_schema as cs
+from typing_extensions import Any, Self, TypeVar
+
+from craft_application import errors
+from craft_application.models import base
+from craft_application.models.constraints import SingleEntryList, UniqueList
+
+
+class Platform(base.CraftBaseModel):
+    """A single platform entry in the platforms dictionary.
+
+    This model defines how a single value under the ``platforms`` key works for a project.
+    """
+
+    build_on: UniqueList[str] | str = pydantic.Field(
+        min_length=1,
+        examples=[
+            "amd64",
+            '["arm64", "riscv64"]',
+        ],
+    )
+    """Architectures to build on.
+
+    If this field is a string containing the build-on architecture, it will be parsed
+    at runtime into a list of strings containing these architectures. This list must
+    contain unique values.
```

Comment:
nit: the first half talks about a singular arch then the second half talks about many archs
```suggestion
    If this field is a string containing the build-on architecture, it will be parsed
    at runtime into a list of strings containing this architecture. This list must
    contain unique values.
```